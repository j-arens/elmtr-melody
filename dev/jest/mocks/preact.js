module.exports = {
    unmountComponent: jest.fn(),
};
