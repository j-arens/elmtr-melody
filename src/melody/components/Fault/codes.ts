export enum ErrorCodes {
    MELODY_GENERIC_FAULT = 1,
    MELODY_NO_TRACKS,
    MELODY_BAD_SOURCE,
}
