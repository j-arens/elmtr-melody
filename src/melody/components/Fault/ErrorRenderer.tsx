import { isEditMode } from '@melody/utils/';
import { h } from 'preact';
import BadSource from './BadSource';
import { ErrorCodes } from './codes';
import DefaultError from './DefaultError';
import FaultContainer from './FaultContainer';
import NoTracksEditor from './NoTracksEditor';
import NoTracksFront from './NoTracksFront';

interface Props {
    errors: Set<ErrorCodes>;
}

export default ({ errors }: Props) => {
    // this check needs to be before the bad source check
    if (errors.has(ErrorCodes.MELODY_NO_TRACKS)) {
        if (isEditMode()) {
            return (
                <FaultContainer allowReload={false}>
                    <NoTracksEditor />
                </FaultContainer>
            );
        }
        return (
            <FaultContainer allowReload={false}>
                <NoTracksFront />
            </FaultContainer>
        );
    }
    if (errors.has(ErrorCodes.MELODY_BAD_SOURCE)) {
        return (
            <FaultContainer allowReload>
                <BadSource />
            </FaultContainer>
        );
    }
    return (
        <FaultContainer allowReload>
            <DefaultError />
        </FaultContainer>
    );
};
