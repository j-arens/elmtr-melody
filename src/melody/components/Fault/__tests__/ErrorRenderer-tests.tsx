import { h } from 'preact';
import { render } from 'preact-render-to-string';
import { ErrorCodes } from '../codes';
import ErrorRenderer from '../ErrorRenderer';

describe('ErrorRenderer', () => {
    it('renders BadSource component if error codes contains bad source error', () => {
        const errors = new Set([ErrorCodes.MELODY_BAD_SOURCE]);
        const component = render(<ErrorRenderer errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('renders NoTracksEditor component if there are no tracks and user is in edit mode', () => {
        global.elementor = {
            isEditMode: jest.fn().mockReturnValue(true),
        };
        const errors = new Set([ErrorCodes.MELODY_NO_TRACKS]);
        const component = render(<ErrorRenderer errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('render NoTracksFront component if there are no tracks and user is not in edit mode', () => {
        global.elementor = {
            isEditMode: jest.fn().mockReturnValue(false),
        };
        const errors = new Set([ErrorCodes.MELODY_NO_TRACKS]);
        const component = render(<ErrorRenderer errors={errors} />);
        expect(component).toMatchSnapshot();
    });

    it('renders DefaultError component for unhandled errors', () => {
        const errors = new Set([9999]);
        const component = render(<ErrorRenderer errors={errors} />);
        expect(component).toMatchSnapshot();
    });
});
