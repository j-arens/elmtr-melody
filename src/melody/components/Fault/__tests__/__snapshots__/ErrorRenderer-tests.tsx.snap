// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorRenderer render NoTracksFront component if there are no tracks and user is not in edit mode 1`] = `
<div class="errorContainer">
  <p class="ErrorMsg" data-cy="error-msg">No tracks are loaded. If you think this an error try refreshing the page,
    otherwise please contact support.</p>
</div>
`;

exports[`ErrorRenderer renders BadSource component if error codes contains bad source error 1`] = `
<div class="errorContainer">
  <p class="ErrorMsg" data-cy="error-msg">Unable to load track, the source was not found.</p>
  <button class="btn errorContainer__reloadBtn"
  role="button">Reload</button>
</div>
`;

exports[`ErrorRenderer renders DefaultError component for unhandled errors 1`] = `
<div class="errorContainer">
  <p class="ErrorMsg">Something went wrong! Please try refreshing the page, otherwise contact
    support.</p>
  <button class="btn errorContainer__reloadBtn" role="button">Reload</button>
</div>
`;

exports[`ErrorRenderer renders NoTracksEditor component if there are no tracks and user is in edit mode 1`] = `
<div class="errorContainer">
  <p class="ErrorMsg" data-cy="error-msg">No tracks are loaded. If you think this an error try refreshing the page,
    otherwise please contact support.</p>
</div>
`;
